userspace/async/async_key_value_source.tpp

```diff
@@ -0,0 +1,288 @@
+		value_type& value,
+		const callback_handler& callback)
+{
+	std::unique_lock<std::mutex> guard(m_mutex);
+
+	if(!m_running && !m_thread.joinable())
+	{
+		m_thread = std::thread(&async_key_value_source::run, this);
+	}
+
+	typename value_map::const_iterator itr = m_value_map.find(key);
+	bool request_complete = (itr != m_value_map.end()) && itr->second.m_available;
+
+	if(!request_complete)
+	{
+		// Haven't made the request yet
+		if (itr == m_value_map.end())
+		{
+			m_value_map[key].m_available = false;
+			m_value_map[key].m_value = value;
+		}
+
+		// Make request to API and let the async thread know about it
+		if (std::find(m_request_set.begin(),
+		              m_request_set.end(),
+		              key) == m_request_set.end())
+		{
+			m_request_queue.push_back(key);
+			m_request_set.insert(key);
+			m_queue_not_empty_condition.notify_one();
+		}
+
+		//
+		// If the client code is willing to wait a short amount of time
+		// to satisfy the request, then wait for the async thread to
+		// pick up the newly-added request and execute it.  If
+		// processing that request takes too much time, then we'll
+		// not be able to return the value information on this call,
+		// and the async thread will continue handling the request so
+		// that it'll be available on the next call.
+		//
+		if (m_max_wait_ms > 0)
+		{
+			m_value_map[key].m_available_condition.wait_for(
+					guard,
+					std::chrono::milliseconds(m_max_wait_ms));
+
+			itr = m_value_map.find(key);
+			request_complete = (itr != m_value_map.end()) && itr->second.m_available;
+		}
+	}
+
+	if(request_complete)
+	{
+		value = itr->second.m_value;
+		m_value_map.erase(key);
+	}
+	else
+	{
+		m_value_map[key].m_callback = callback;
+	}
+
+	return request_complete;
+}
+
+template<typename key_type, typename value_type>
+bool async_key_value_source<key_type, value_type>::dequeue_next_key(key_type& key)
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
+	bool key_found = false;
+
+	if(m_request_queue.size() > 0)
+	{
+		key_found = true;
+
+		key = m_request_queue.front();
+		m_request_queue.pop_front();
+		m_request_set.erase(key);
+	}
+
+	return key_found;
+}
+
+template<typename key_type, typename value_type>
+value_type async_key_value_source<key_type, value_type>::get_value(
+		const key_type& key)
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
+
+	return m_value_map[key].m_value;
+}
+
+template<typename key_type, typename value_type>
+void async_key_value_source<key_type, value_type>::store_value(
+		const key_type& key,
+		const value_type& value)
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
+
+	if (m_value_map[key].m_callback)
+	{
+		m_value_map[key].m_callback(key, value);
+		m_value_map.erase(key);
+	}
+	else
+	{
+		m_value_map[key].m_value = value;
+		m_value_map[key].m_available = true;
+		m_value_map[key].m_available_condition.notify_one();
+	}
+}
+
+/**
+ * Prune any "old" outstanding requests.  This method expect that the caller
+ * is holding m_mutex.
```

Comment:
can we enforce this somehow? e.g. by passing a std::lock_guard as a parameter

Reply:
I don't know how I feel about passing in an object that this method won't use.  If this were a public API, I'd be more concerned about it, but since this is a private method I think documenting the precondition should be enough.  Does that seem legit?